### PR TITLE
Menus on stream deck

### LINKED ISSUE #47: Support Adding Encounters Cards to Player Hands on StreamDeck

### PR DESCRIPTION
**Pull request by slyjeff**
Long holding on an encounter card will now allow you to put it in an investigator's hand.
(Eventually we should probably only do this for hidden cards, as that's the only case where this is an option).

Part of this was refactoring how the ShowCardList action is initialized, as it is drawing from the same source data as resolving these menu items (card groups).

Fixed some buggy/inconsitent behavior in how data was initialize being sent down to the streamdeck and the connection was being maintained

### CODE REVIEW

File: ArkhamOverlay/Services/BroadcastService.cs [hunk 1]
```diff
@@ -5,51 +5,78 @@
...
-                        //this clearly is not cool- stop trying to talk
-                        _logger.LogException(ex, $"Error sending status to port {port}.");
-                        portsToRemove.Add(port);
                     }
-                }
 
-                foreach (var portToRemove in portsToRemove) {
-                    Ports.Remove(portToRemove);
+                    foreach (var portToRemove in portsToRemove) {
+                        _ports.Remove(portToRemove);
```
Comment (davilimap) [inline ArkhamOverlay/Services/BroadcastService.cs:78]: Should this be in a lock?

File: StreamDeckPlugin/Actions/DynamicAction.cs [hunk 2]
```diff
@@ -125,10 +127,19 @@ public class DynamicAction : StreamDeckAction<ActionWithDeckSettings>, IButtonCo
 
         private object _keyUpLock = new object();
         private bool _keyIsDown = false;
+
         protected override Task OnKeyDown(ActionEventArgs<KeyPayload> args) {
             _settings = args.Payload.GetSettings<ActionWithDeckSettings>();
-            _keyIsDown = true;
 
+            if (_dynamicActionOption != null) {
+                //we are showing a menyu item, so alert the dynamic action manager instead of our normal behavior
```
Comment (davilimap) [inline StreamDeckPlugin/Actions/DynamicAction.cs]: nit: menu

File: StreamDeckPlugin/Actions/TrackStatAction.cs [hunk 3]
```diff
@@ -102,8 +102,12 @@ public abstract class TrackStatAction : StreamDeckAction {
         }
 
         public void UpdateValue(int value) {
-            _value = value;
-            SetTitleAsync(_value.ToString());
+            try {
+                _value = value;
+                SetTitleAsync(_value.ToString());
+            } catch {
+                //sometimes this happens before we are set up, and the interals throw an exception
```
Comment (davilimap) [inline StreamDeckPlugin/Actions/TrackStatAction.cs]: nit: internals

File: StreamDeckPlugin/Services/CardGroupStore.cs [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+using ArkhamOverlay.Common.Enums;
+using ArkhamOverlay.Common.Services;
+using ArkhamOverlay.Events;
+using StreamDeckPlugin.Events;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace StreamDeckPlugin.Services {
+    /// <summary>
+    /// Stores information foreach card group
```
Comment (davilimap) [inline StreamDeckPlugin/Services/CardGroupStore.cs:11]: nit: space in for each

File: StreamDeckPlugin/Services/CardGroupStore.cs [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+        /// <summary>
+        /// Retrieve Card group Info from the store
+        /// </summary>
+        /// <param name="cardGroupId">Id of the card group info to retrieve</param>
+        /// <returns>Card group info</returns>
+        ICardGroupInfo GetCardGroupInfo(CardGroupId cardGroupId);
+    }
+
+    /// <summary>
+    /// Stores information foreach card group
```
Comment (davilimap) [inline StreamDeckPlugin/Services/CardGroupStore.cs]: nit: space

File: StreamDeckPlugin/Services/CardGroupStore.cs [hunk 6]
```diff
@@ -0,0 +1,90 @@
...
+            _eventBus.PublishStreamDeckCardGroupInfoChanged(eventData);
+
+            UpdateCardGroupInfo(eventData);            
+        }
+
+        /// <summary>
+        /// Add or update the data about this card group in the store
+        /// </summary>
+        /// <param name="cardGroupInfo">Info about the card group</param>
+        public void UpdateCardGroupInfo(ICardGroupInfo cardGroupInfo) {
```
Comment (davilimap) [inline StreamDeckPlugin/Services/CardGroupStore.cs]: Why is this empty?

File: StreamDeckPlugin/Services/ImageService.cs [hunk 7]
```diff
@@ -7,9 +9,19 @@
...
         void UpdateButtonImage(string name, byte[] bytes);
+
+        /// <summary>
+        /// Check to see if this image is stored- if it is not, retrieve it
+        /// </summary>
+        /// <param name="imageId">Id of the image in the cache- used to store and retrieve it</param>
+        /// <param name="cardGroupId">Card Group for the image</param>
+        /// <param name="buttonMode">ButtonMode for the image- may be null if the image is for the Card Group </param>
+        /// <param name="index">Index of the image- may be null if the image is for the Card Group</param>
+        void LoadImage(string imageId, CardGroupId cardGroupId, ButtonMode? buttonMode = null, int? index = null);
```
Comment (davilimap) [inline StreamDeckPlugin/Services/ImageService.cs:52]: Minor: for consistency, add comments to the rest of the interface

File: StreamDeckPlugin/Actions/DynamicAction.cs (hunk 2 above)
Comment (slyjeff) [reply StreamDeckPlugin/Actions/DynamicAction.cs]: fixed

File: StreamDeckPlugin/Actions/TrackStatAction.cs (hunk 3 above)
Comment (slyjeff) [reply StreamDeckPlugin/Actions/TrackStatAction.cs]: fixed

File: StreamDeckPlugin/Services/CardGroupStore.cs (hunk 5 above)
Comment (slyjeff) [reply StreamDeckPlugin/Services/CardGroupStore.cs]: fixed

File: StreamDeckPlugin/Services/CardGroupStore.cs (hunk 6 above)
Comment (slyjeff) [reply StreamDeckPlugin/Services/CardGroupStore.cs]: ooops- this was a method that used to be called/used when there were different ways to update this store, but that funcationilty has gone away and I didn't completely clean it up

File: StreamDeckPlugin/Services/ImageService.cs (hunk 7 above)
Comment (slyjeff) [reply StreamDeckPlugin/Services/ImageService.cs:52]: Done- getting this into shape slowly :D

**Review by davilimap — APPROVED** [review]

**Pull request merged**